cpp/src/sampling/detail/sampling_utils_impl.cuh

```diff
@@ -0,0 +1,886 @@
+       id_end               = id_end.data(),
+       id_seg_count         = id_begin.size(),
+       vertex_count_offsets = vertex_count_offsets.data(),
+       glbl_degree_offsets  = global_degree_offsets.data(),
+       majors               = majors.data(),
+       minors               = minors.data(),
+       partitions           = partitions.data(),
+       hypersparse_begin    = hypersparse_begin.data(),
+       invalid_vertex_id,
+       indices_per_major] __device__(auto index) {
+        // major which this edge index refers to
+        auto loc      = index / indices_per_major;
+        auto major    = active_majors[loc];
+        majors[index] = major;
+
+        // Find which partition id did the major belong to
+        auto partition_id = thrust::distance(
+          id_end, thrust::upper_bound(thrust::seq, id_end, id_end + id_seg_count, major));
+        // starting position of the segment within global_degree_offset
+        // where the information for partition (partition_id) starts
+        //  vertex_count_offsets[partition_id]
+        // The relative location of offset information for vertex id v within
+        // the segment
+        //  v - seg[partition_id]
+        vertex_t location_in_segment;
+        if (major < hypersparse_begin[partition_id]) {
+          location_in_segment = major - id_begin[partition_id];
+        } else {
+          auto row_hypersparse_idx =
+            partitions[partition_id].major_hypersparse_idx_from_major_nocheck(major);
+          if (row_hypersparse_idx) {
+            location_in_segment = *(row_hypersparse_idx)-id_begin[partition_id];
+          } else {
+            minors[index] = invalid_vertex_id;
+            return;
+          }
+        }
+
+        // csr offset value for vertex v that belongs to partition (partition_id)
+        auto offset_ptr                = partitions[partition_id].offsets();
+        auto sparse_offset             = offset_ptr[location_in_segment];
+        auto local_out_degree          = offset_ptr[location_in_segment + 1] - sparse_offset;
+        vertex_t const* adjacency_list = partitions[partition_id].indices() + sparse_offset;
+        // read location of global_degree_offset needs to take into account the
+        // partition offsets because it is a concatenation of all the offsets
+        // across all partitions
+        auto location    = location_in_segment + vertex_count_offsets[partition_id];
+        auto g_dst_index = edge_index_first[index];
+        if (g_dst_index >= 0) {
+          minors[index]           = adjacency_list[g_dst_index];
+          edge_index_first[index] = g_dst_index;
+        } else {
+          minors[index] = invalid_vertex_id;
+        }
+      });
+  }
+
+  if (weights) {
+    auto input_iter = thrust::make_zip_iterator(
+      thrust::make_tuple(majors.begin(), minors.begin(), weights->begin()));
+
+    auto compacted_length = thrust::distance(
+      input_iter,
+      thrust::remove_if(
```

Comment:
FYI: The current version of thrust::remove_if does not work properly if minors.size() overflows 32bit integer.

Reply:
Added FIXME to both of these remove_if calls in this file (both branches of the if)